### PR TITLE
Fixed translate bug for articles translation.

### PR DESCRIPTION
**Pull request by abhiprojectz**
**Bug**:
`Translate content` gives server error 500 when translating for articles.

**Description:**
This fixes the above bug on `/dashboard/translate/answers` page.

### CODE REVIEW

**Review by badrihippo — CHANGES_REQUESTED** [review]
This works, logically speaking, but better not to make such changes in the base views because over time it'll lead to spaghetti code (one code trying to account for too many cases, like what's happened to the Search page right now).

We already have separate functions to account for two different cases (`CreateArticleTranslation` for articles and `CreateAnswerTranslation` for answers). So instead of putting the `if` condition here, you should create a `get_form` function under `CreateAnswerTranslation` which inherits from the parent one; something like

```
class CreateAnswerTranslation(BaseStartTranslation):
    ...

    def get_form(self):
        form = super().get_form()

        available_languages = self.answer.list_available_languages()

        unavailable_languages = []
        ...
```

Then in the `BaseStartTranslation` use `self.source` instead of `self.answer`. The reason is that `source` is the *normal* way to use a translation; answers are a *special case* where you get the language from the answer instead of the `source` (question).

The reason I'm suggesting all this is that, when someone looks at the code later, they can easily read `BaseStartTranslation` to see what it does, and then they can see that `CreateAnswerTranslation` is customising it in some way (getting the language list from the answer instead of the question). This is easier to understand than having to understand one large code that accounts for all the cases. It may seem like a bit more work right now, but in the long run it makes the code much easier to maintain.

Actually even the current case is a good example. Somebody tried to customise `BaseStartTranslation` directly for answers, but that ended up affecting the other translation views because it was looking for `answers` instead of looking for `source`. If they had customised `CreateAnswerTranslation` instead of `BaseStartTranslation` this bug would never have come up (or at least it would have been very easy to track down). By the way, I don't know who changed that line; it could have even been me, but whoever did it that wasn't a good idea!

Hope that was clear; let me know if you have any doubts or questions 🙂

**Comment by badrihippo** [comment]
@abhiprojectz just to clarify, what you did was not *wrong*: it did solve the problem at hand. But design-wise, since this is such a large project, there are better ways to implement the solution that will work out better in the long run!

File: dashboard/views.py [hunk 1]
```diff
@@ -1603,8 +1603,17 @@ def get_source(self, source):
     def get_form(self):
         form = super().get_form()
 
+        # get full absolute request url 
+        absolute_url = self.request.get_full_path()
+
+        # check whether its article
```
Comment (pskeshu) [inline dashboard/views.py:1609]: typo

**Comment by pskeshu** [comment]
>If they had customised CreateAnswerTranslation instead of BaseStartTranslation this bug would never have come up (

Is this a larger problem with the library API?

**Comment by abhiprojectz** [comment]
@badrihippo  @pskeshu   please suggest a better way to implement the above change in templates like in `translation/base_edit.html`  template 
```
...
 <select id="lang_from" name="lang_from" class="input-plain text-center">
          {% for value, label in object.source.list_available_languages %}
          <option value="{{ value }}"{% if value == object.source.tr_language %} selected{% endif %}>{{ label }}</option>
          {% endfor %}
        </select>
...
```

hghgg

the languages filter (lang_from,lang_to) are being set from source (object) language on `translation interface` and same is used for both articles and answers (resources or research etc) so without making changes in `translation/base_edit.html`   template and without using `{% if/else %}` ,
how to extend this change to `answer_edit` and `article_edit`  template such that `base_edit` have some common fields after the `lang filters`.

**Comment by badrihippo** [comment]
> > If they had customised CreateAnswerTranslation instead of BaseStartTranslation this bug would never have come up (
> 
> Is this a larger problem with the library API? @badrihippo

@pskeshu not a *problem* per se, but a usage pattern that needs to be encouraged. Perhaps it needs to be documented a bit better; on the other hand once 2-3 exceptions like this are added the recommended structure will hopefully become more apparent to new programmers

**Comment by badrihippo** [comment]
@abhiprojectz two possible options; choose whichever you're more comfortable with.

### Option 1: Template-level override

You can create a new `{% block %}` like `{% block language_selector %}`. The `base_edit.html` can have the default value, i.e.

```
        ...
        {% block language_selector %}
        <select id="lang_from" name="lang_from" class="input-plain text-center">
          {% for value, label in object.source.list_available_languages %}
          <option value="{{ value }}"{% if value == object.source.tr_language %} selected{% endif %}>{{ label }}</option>
          {% endfor %}
        </select>
        {% block language_selector %}
        ...
```

The no change will be necessary in `article_edit.htm`, but in `answer_edit` you can override by putting a different `{% block language-selector %}` there, similarly to how it is done for `{% block title %}`, etc. in other places.

### Option 2: View-level override

Alternatively, you can set a new context variable called `available_languages` in `EditAnswerTranslation` in `views.py`. Then in `base.html` you can change the line to `{% for value, label in available_languages or object.source.list_available_languages %}` (we are keeping the `or` so people don't have to manually compute the `available_languages` when sticking with the default option).

**Comment by abhiprojectz** [comment]
Thanks @badrihippo  ,made the changes.

**Review by badrihippo — APPROVED** [review]
Looking good!

**Pull request merged**